File: real-time-risk/go/riskreadmodels/store.go
// Package riskreadmodels is the Risk Read Models service (CQRS off DynamoDB Streams, reference doc
// §3): it consumes the Trade Ledger's DynamoDB Stream in the background and projects it into a
// per-book, per-symbol net-position + realized-cash view, served by GET /books/{book}/positions.
// It is the Go port of real-time-risk/dotnet/RiskReadModels.
package riskreadmodels

import (
	"sort"
	"sync"

	"github.com/daniellepelley/benzene-patterns/real-time-risk/go/contracts"
)

// BookPositionsStore is the projection this service builds: per book, per symbol net position +
// realized cash, folded from TradeBooked events as the projector consumes them off the ledger's
// DynamoDB Stream. In-memory only for this first slice (resets on restart/redeploy) - a real store
// isn't needed to prove the pattern yet. Mirrors .NET's BookPositionsStore.
//
// A single mutex guards all books. That is deliberately coarse: the write side is one background
// poller applying records serially, and the read side is a cheap snapshot copy, so contention is a
// non-issue at this demo's scale and a per-book lock would be premature.
type BookPositionsStore struct {
	mu    sync.Mutex
	books map[string]*book
}

type book struct {
	positions map[string]contracts.PositionView
	// applied is the set of stream versions already folded in, so a redelivered record is a no-op.
	applied                 map[int64]struct{}
	projectedThroughVersion int64
}

// NewBookPositionsStore returns an empty projection store.
func NewBookPositionsStore() *BookPositionsStore {
	return &BookPositionsStore{books: make(map[string]*book)}
}

// Apply folds one TradeBooked event into the book's projection. It is idempotent by
// (book, version): DynamoDB Streams is at-least-once and the projector re-reads from TRIM_HORIZON
// on restart, so a redelivered record must not be double-applied. The math mirrors .NET exactly:
//
//	signedQuantity = Buy ? +q : -q
//	cashFlow       = Sell ? +(q*p) : -(q*p)
func (s *BookPositionsStore) Apply(trade contracts.TradeBooked, version int64) {
	s.mu.Lock()
	defer s.mu.Unlock()

	b, ok := s.books[trade.Book]
	if !ok {
		b = &book{positions: make(map[string]contracts.PositionView), applied: make(map[int64]struct{})}
		s.books[trade.Book] = b
	}
	if _, seen := b.applied[version]; seen {
		return
	}
	b.applied[version] = struct{}{}

	current := b.positions[trade.Symbol]
	signedQuantity := trade.Quantity
	if trade.Side != contracts.Buy {
		signedQuantity = -trade.Quantity
	}
	cashFlow := -(trade.Quantity * trade.Price)
	if trade.Side == contracts.Sell {
		cashFlow = trade.Quantity * trade.Price
	}

	b.positions[trade.Symbol] = contracts.PositionView{
		Symbol:       trade.Symbol,
		NetQuantity:  current.NetQuantity + signedQuantity,
		RealizedCash: current.RealizedCash + cashFlow,
	}

	if version > b.projectedThroughVersion {
		b.projectedThroughVersion = version
	}
}

// Query reads a book's current projection, positions sorted by symbol. An unknown book returns an
// empty (zero-trade) result, matching .NET.
func (s *BookPositionsStore) Query(bookID string) contracts.BookPositionsResponse {
	s.mu.Lock()
	defer s.mu.Unlock()

	b, ok := s.books[bookID]
	if !ok {
		return contracts.BookPositionsResponse{Book: bookID, Positions: []contracts.PositionView{}}
	}

	positions := make([]contracts.PositionView, 0, len(b.positions))
	for _, p := range b.positions {
		positions = append(positions, p)
	}
	sort.Slice(positions, func(i, j int) bool { return positions[i].Symbol < positions[j].Symbol })

	return contracts.BookPositionsResponse{
		Book:                    bookID,
		Positions:               positions,
		ProjectedThroughVersion: b.projectedThroughVersion,
	}
}

File: real-time-risk/python/risk_read_models/store.py
"""The read-model projection — per book, per symbol net position + realized cash.

Mirrors .NET ``RiskReadModels/BookPositionsStore.cs``. Folded from :class:`TradeBooked` events as the
:class:`TradeStreamProjector` consumes them off the ledger's DynamoDB Stream. In-memory only for this
first slice (it resets on restart/redeploy) — a real store isn't needed to prove the pattern yet.

**Idempotent by (book, version).** DynamoDB Streams is at-least-once, and the local poller re-reads
from ``TRIM_HORIZON`` on every restart, so a redelivered version must not be double-applied. Each
book tracks the set of versions it has already folded and ignores repeats. A ``threading.Lock`` guards
mutation because the projector runs on its own thread (``asyncio.to_thread``) while HTTP queries read
concurrently on the event loop.
"""

from __future__ import annotations

import threading

from contracts import BookPositionsResponse, PositionView, TradeBooked, TradeSide


class _Book:
    __slots__ = ("positions", "applied_versions", "projected_through_version")

    def __init__(self) -> None:
        self.positions: dict[str, PositionView] = {}
        self.applied_versions: set[int] = set()
        self.projected_through_version: int = 0


class BookPositionsStore:
    """The queryable per-book projection built from the ledger's events."""

    def __init__(self) -> None:
        self._books: dict[str, _Book] = {}
        self._gate = threading.Lock()

    def apply(self, trade: TradeBooked, version: int) -> None:
        """Fold one :class:`TradeBooked` into its book's projection. Idempotent by (book, version)."""
        with self._gate:
            book = self._books.setdefault(trade.book, _Book())
            if version in book.applied_versions:
                return
            book.applied_versions.add(version)

            current = book.positions.get(trade.symbol, PositionView(symbol=trade.symbol))
            # Buy adds to the net position and spends cash; sell subtracts and brings cash in.
            signed_quantity = (
                trade.quantity if trade.side == TradeSide.BUY else -trade.quantity
            )
            cash_flow = (
                trade.quantity * trade.price
                if trade.side == TradeSide.SELL
                else -(trade.quantity * trade.price)
            )
            book.positions[trade.symbol] = PositionView(
                symbol=trade.symbol,
                net_quantity=current.net_quantity + signed_quantity,
                realized_cash=current.realized_cash + cash_flow,
            )

            if version > book.projected_through_version:
                book.projected_through_version = version

    def query(self, book_id: str) -> BookPositionsResponse:
        """Read a book's current projection; an unknown book returns an empty (zero-trade) result."""
        with self._gate:
            book = self._books.get(book_id)
            if book is None:
                return BookPositionsResponse(book=book_id, positions=[], projected_through_version=0)
            return BookPositionsResponse(
                book=book_id,
                positions=[book.positions[s] for s in sorted(book.positions)],
                projected_through_version=book.projected_through_version,
            )
